File: dart/lib/src/models/link_stats.dart
//
//  rerout
//  link_stats.dart
//
//  Official Dart SDK for the Rerout branded-link API.
//
//  Created by Ngonidzashe Mangudya on 20/05/2026.
//  Copyright (c) 2026 Codecraft Solutions
//  Licensed under the MIT License
//  https://codecraftsolutions.co.za
//

import 'package:meta/meta.dart';
import 'package:rerout/src/models/stats_breakdown.dart';

/// Analytics for a single short link across the requested window.
@immutable
class LinkStats {
  /// Creates [LinkStats].
  const LinkStats({
    required this.code,
    required this.days,
    required this.totalClicks,
    required this.qrScans,
    required this.countries,
    required this.referrers,
  });

  /// Parses [LinkStats] from the API JSON envelope.
  factory LinkStats.fromJson(Map<String, dynamic> json) => LinkStats(
    code: json['code'] as String? ?? '',
    days: (json['days'] as num?)?.toInt() ?? 0,
    totalClicks: (json['total_clicks'] as num?)?.toInt() ?? 0,
    qrScans: (json['qr_scans'] as num?)?.toInt() ?? 0,
    countries:
        (json['countries'] as List<dynamic>? ?? const <dynamic>[])
            .whereType<Map<String, dynamic>>()
            .map(StatsBreakdown.fromJson)
            .toList(growable: false),
    referrers:
        (json['referrers'] as List<dynamic>? ?? const <dynamic>[])
            .whereType<Map<String, dynamic>>()
            .map(StatsBreakdown.fromJson)
            .toList(growable: false),
  );

  /// The short code these stats belong to.
  final String code;

  /// Window size in days the totals span.
  final int days;

  /// Total clicks in the window.
  final int totalClicks;

  /// Subset of [totalClicks] attributed to a QR scan.
  final int qrScans;

  /// Top countries.
  final List<StatsBreakdown> countries;

  /// Top referrers.
  final List<StatsBreakdown> referrers;
}

File: php/src/Models/LinkStats.php
<?php

declare(strict_types=1);

namespace Rerout\Models;

use InvalidArgumentException;

/**
 * Per-link click stats over a configurable window.
 */
final readonly class LinkStats
{
    /**
     * @param list<StatsBreakdown> $countries
     * @param list<StatsBreakdown> $referrers
     */
    public function __construct(
        public string $code,
        public int $days,
        public int $totalClicks,
        public int $qrScans,
        public array $countries,
        public array $referrers,
    ) {
    }

    /**
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        $code = $data['code'] ?? null;
        $days = $data['days'] ?? null;
        $totalClicks = $data['total_clicks'] ?? null;
        $qrScans = $data['qr_scans'] ?? null;

        if (!is_string($code)) {
            throw new InvalidArgumentException("Expected string field 'code' in LinkStats payload.");
        }
        if (!is_int($days) || !is_int($totalClicks) || !is_int($qrScans)) {
            throw new InvalidArgumentException('Expected int fields days/total_clicks/qr_scans in LinkStats payload.');
        }

        return new self(
            code: $code,
            days: $days,
            totalClicks: $totalClicks,
            qrScans: $qrScans,
            countries: self::mapBreakdowns($data['countries'] ?? []),
            referrers: self::mapBreakdowns($data['referrers'] ?? []),
        );
    }

    /**
     * @param mixed $raw
     *
     * @return list<StatsBreakdown>
     */
    private static function mapBreakdowns(mixed $raw): array
    {
        if (!is_array($raw)) {
            return [];
        }
        $out = [];
        foreach ($raw as $row) {
            if (is_array($row)) {
                /** @var array<string, mixed> $row */
                $out[] = StatsBreakdown::fromArray($row);
            }
        }

        return $out;
    }
}
